neo4j/driver/Neo4jTransaction.java

```diff
@@ -27,29 +27,57 @@
 
 import java.util.List;
 
-public class Neo4jTransaction implements Transaction {
+public abstract class Neo4jTransaction implements Transaction {
 
     private static final Logger LOG = LoggerFactory.getLogger(Neo4jTransaction.class);
 
-    private final Session session;
+    protected final Session session;
 
     public Neo4jTransaction(Session session) {
         // TODO: why are we not passing a Transaction in here?
         this.session = session;
     }
 
+    static class Read extends Neo4jTransaction {
+
+        public Read(Session session) {
+            super(session);
+        }
+
+        public List<Record> execute(Query query) {
+            return session.readTransaction(tx -> tx.run(query).list());
+        }
+
+        @Override
+        public void commit() {
+            throw new UnsupportedOperationException("There should not be a call to .commit() for read transactions.");
+        }
+    }
+
+    static class Write extends Neo4jTransaction {
+
+        public Write(Session session) {
+            super(session);
+        }
+
+        public List<Record> execute(Query query) {
+            return session.writeTransaction(tx -> tx.run(query).list());
+        }
+
+        /**
+         * Not necessary when using Neo4j's Transaction Functions
+         */
+        @Override
+        public void commit() {}
+    }
+
     public List<Record> execute(Query query) {
         return session.writeTransaction(tx -> {
             Result result = tx.run(query);
             return result.list();
         });
     }
```

Comment:
Does this method still needs to exist now that the class is always extended?

Reply:
I don't understand, this method is only present in the base interface `Transaction` and in the `Neo4jTransaction.Write` and `Neo4jTransaction.Read`, not in `Neo4jTransaction`. I've reordered the methods to be above the inner classes in `Neo4jTransaction` though as they were out of order.